lib/harmony/resources/storagemgr/storagemgt.go

```diff
@@ -0,0 +1,251 @@
+// Storagemgr is a utility for harmonytask tasks
+// to manage the available space.
+// Ex:
+//
+//	TaskTypeDefinition{
+//	   Storage: storMgr.MakeFuncs(UsageTemporary, 1<<30, func(taskID int) string {
+//					return fmt.Sprintf("task-%d", taskID)   // or use SQL to get the sector ID
+//			}),
+//	}
+//
+// Also useful: s.CombineFuncs(s1, s2, s3)
+package storagemgr
+
+import (
+	"bytes"
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	logging "github.com/ipfs/go-log/v2"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/lib/harmony/resources"
+	"github.com/filecoin-project/lotus/storage/sealer/storiface"
+)
+
+var log = logging.Logger("harmonytask")
+
+type path string
+
+var claimsMx sync.Mutex
+var claims = map[path][]consumption{}
+
+var purposes = map[Usage][]path{}
+
+type Usage int
+
+const (
+	UsagePermanent = iota
+	UsageTemporary
+	UsageCache
+	UsageStaging
+)
+
+type StorageMgr struct { // Functions are attached here to force users to call New()
+}
+
+func New(paths []storiface.StoragePath) *StorageMgr {
+	// Note: this may be called multiple times per process.
+	// populate purposes (global)
+	for _, p := range paths {
+		// @reviewer help this logic make sense
+		var t Usage = UsagePermanent
+		if p.CanSeal {
+			t = UsageStaging
+		}
+		if strings.Contains(p.LocalPath, "cache") {
+			t = UsageCache
+		}
+		if strings.Contains(p.LocalPath, "tmp") {
+			t = UsageTemporary
+		}
+		purposes[t] = append(purposes[t], path(p.LocalPath))
+	}
+
+	return &StorageMgr{}
```

Comment:
https://github.com/filecoin-project/lotus/tree/master/storage/sealer#package-paths has some rough description of how the storage subsystem roughly works.

Basically there can be multiple storage paths attached to a given node, and each storage path can hold one of 5 "file types" (sealed data, unsealed data, etc.)

Then each of those paths can say whether it wants to support Sealing on Long Term storage (neither and both are valid combinations)

Here we're mostly reimplementing what exists in https://github.com/filecoin-project/lotus/blob/master/storage/paths/local.go#L402, and in general what the Local paths subsystem is doing, and I'd really rather not do that given how much battle testing that part of the code has seen - e.g. there are no syscalls in the calls in the calls relevant to scheduling, because those will hang (e.g. try to start 10 32G mv-s on an NFS share, then see `df -h` hang; dying but not dead yet ssds/hdds are also similar)

The existing paths.Local Reserve mechanism also is going to send the reservation stats to the sector index, which then will show that information in the storage cli / prometheus / will be query-able from postgres;

Both impls are missing one thing - paths can be shared between nodes - e.g. one can have one GPU box with 10 GPUs, one raided nvme mount - then neither of the systems will be able to account properly for reserved storage in the shared mount because there are 10 processes with independent in-memory accounting.

(but now that we have the db storage index it should be reasonably easy to change paths.Local to do the reservation accounting through the database)